### PR TITLE
rust-session-manager: handle `crate_name` attribute

### LINKED ISSUE #789: Default Crate Name is wrong
The default in rustc is the source filename without `.rs` or `rust_out` if the source is stdin.

_Originally posted by @bjorn3 in https://github.com/Rust-GCC/gccrs/pull/788#discussion_r740758707_

### PR DESCRIPTION
**Pull request by liushuyu**
- rust-session-manager: handle `crate_name` attribute

Fix #789 properly

### CODE REVIEW

**Review by CohenArthur — APPROVED** [review]
LGTM

**Review by philberty — COMMENTED** [review]
I hate to be pedantic but its worth discussing the crate_name_set_manually flag and the crate_name_changed flags.

It seems that crate_name_changed flag is set to false but never changed so it doesn't really have an affect.. Then the crate_name_set_manually seems to be that it might be better to have a helper function on the options struct for:

```
bool is_crate_name_set() const { return !crate_name.empty(); }
```

It seems that the place where the crate_name_set_manually flag is used would be better served by a is_crate_name_set helper which seems more readable to me.

Please correct me if I am missing something.

File: gcc/rust/rust-session-manager.cc [hunk 1]
```diff
@@ -582,6 +573,56 @@ Session::parse_files (int num_files, const char **files)
...
+	  rust_error_at (attr.get_locus (),
+			 "%<-frust-crate-name%> and %<#[crate_name]%> are "
+			 "required to match, but %qs does not match %qs",
+			 options.crate_name.c_str (), msg_str.c_str ());
+	}
+      options.set_crate_name (msg_str);
+      mappings->set_crate_name (mappings->get_current_crate (), msg_str);
+    }
+
+  options.crate_name_set_manually |= crate_name_changed;
```
[the remark this answers is not in the document]
Comment (liushuyu) [reply gcc/rust/rust-session-manager.cc:619]: > Do we need crate_name_changed? It looks like its always false here.

Sorry, I forgot to add the logic to flip it to `true`. It's now corrected.

**Comment by liushuyu** [comment]
> I hate to be pedantic but its worth discussing the crate_name_set_manually flag and the crate_name_changed flags.
> 
> It seems that crate_name_changed flag is set to false but never changed so it doesn't really have an affect.. Then the crate_name_set_manually seems to be that it might be better to have a helper function on the options struct for:
> 
> ```
> bool is_crate_name_set() const { return !crate_name.empty(); }
> ```
> 
> It seems that the place where the crate_name_set_manually flag is used would be better served by a is_crate_name_set helper which seems more readable to me.
> 
> Please correct me if I am missing something.

I forgot to add a comment to describe where this variable would be useful. Now it's corrected.

The idea is that inferred crate name will be used as a stand-in to avoid a crate not having any name attached to it prior to `parse_file`.

**Review by philberty — APPROVED** [review]
Thnks for updating the logic, LGTM

**Pull request merged**